File: Haskell/FP-2-uzd/src/MExpression/Parse.hs
module MExpression.Parse where

import Types
import Data.Bool

parse :: String -> [Move]
parse str = 
    let
        str1 = removeWhiteSpaces str
        str2 = removeMExpressionPrefix str1
        str3 = removeOpeningSquareBracket str2
        (moves, result) = parseMExpressions [] str3
    in reverse moves

parseMExpressions :: [Move] -> String -> ([Move], String)
parseMExpressions acc (']':t) = (acc, t)
parseMExpressions acc (';':t) = parseMExpressions acc t
parseMExpressions acc mExpressions =
    let
        (propertyName1, mExpressions1) = readPropertyName mExpressions
        mExpressions2 = removeSemicolon mExpressions1
        (move, rest) = parseMExpression mExpressions2
    in
        parseMExpressions (move:acc) rest

parseMExpression :: String -> (Move, String)
parseMExpression mExpression =
    let
       mExpression1 = removeMExpressionPrefix mExpression
       mExpression2 = removeOpeningSquareBracket mExpression1
       
       (propertyName1, mExpression3) = readPropertyName mExpression2
       mExpression4 = removeSemicolon mExpression3
       
       (property1, mExpression5) = readProperty mExpression4
       mExpression6 = removeSemicolon mExpression5
       
       (propertyName2, mExpression7) = readPropertyName mExpression6
       mExpression8 = removeSemicolon mExpression7
       (property2, mExpression9) = readProperty mExpression8
       mExpression10 = removeSemicolon mExpression9
       
       (propertyName3, mExpression11) = readPropertyName mExpression10
       mExpression12 = removeSemicolon mExpression11
       (property3, mExpression13) = readProperty mExpression12
       mExpression14 = removeClosingSquareBracket mExpression13
       
       rest = mExpression14
       
       x = if (propertyName1 == 'x')
         then property1
         else if (propertyName2 == 'x')
           then property2
           else property3
       y = if (propertyName1 == 'y')
         then property1
         else if (propertyName2 == 'y')
           then property2
           else property3
       v = if (propertyName1 == 'v')
         then property1
         else if (propertyName2 == 'v')
           then property2
           else property3
       trueV = if (v == 3)
         then 'x'
         else 'o'
    in
        (Move x y trueV, rest)

removeWhiteSpaces :: [Char] -> [Char]
removeWhiteSpaces str =
    removeWhiteSpaces' [] str
    where
        removeWhiteSpaces' :: [Char] -> [Char] -> [Char]
        removeWhiteSpaces' acc [] = acc
        removeWhiteSpaces' acc (h:t) =
            if (h /= ' ')
                then removeWhiteSpaces' (acc++[h]) t
                else removeWhiteSpaces' acc t

removeOpeningSquareBracket :: String -> String
removeOpeningSquareBracket ('[':t) = t
removeOpeningSquareBracket _ = error "'[' expected"

removeClosingSquareBracket :: String -> String
removeClosingSquareBracket (']':t) = t
removeClosingSquareBracket _ = error "']' expected"

removeMExpressionPrefix :: String -> String
removeMExpressionPrefix ('m':t) = t
removeMExpressionPrefix _ = error "'m' expected"

removeSemicolon :: String -> String
removeSemicolon (';':t) = t
removeSemicolon _ = error "';' expected"

readPropertyName :: String -> (Char, String)
--readPropertyName ('\"':'x':'\"':t) = ('x', t)
--readPropertyName ('\"':'X':'\"':t) = ('x', t)
--readPropertyName ('\"':'y':'\"':t) = ('y', t)
--readPropertyName ('\"':'Y':'\"':t) = ('y', t)
--readPropertyName ('\"':'v':'\"':t) = ('v', t)
--readPropertyName ('\"':'V':'\"':t) = ('v', t)
readPropertyName ('\"':x:'\"':t) = (x, t)
readPropertyName _  = error "Property name expected"

--readDigit :: String -> (Int, String)
--readDigit ('0':rest) = (0, rest)
--readDigit ('1':rest) = (1, rest) 
--readDigit ('2':rest) = (2, rest) 
--readDigit _ = error "Digit expected"
--
--readPlayer :: String -> (Char, String)
--readPlayer ('\"': 'x' : '\"': rest) = ('x', rest)
--readPlayer ('\"': 'X' : '\"': rest) = ('x', rest)
--readPlayer ('\"': 'o' : '\"': rest) = ('o', rest)
--readPlayer ('\"': 'O' : '\"': rest) = ('o', rest)
--readPlayer _ = error "Player expected"

readProperty :: String -> (Int, String)
readProperty ('0':rest) = (0, rest)
readProperty ('1':rest) = (1, rest)
readProperty ('2':rest) = (2, rest)
readProperty ('\"': 'x' : '\"': rest) = (3, rest)
readProperty ('\"': 'X' : '\"': rest) = (3, rest)
readProperty ('\"': 'o' : '\"': rest) = (4, rest)
readProperty ('\"': 'O' : '\"': rest) = (4, rest)
readProperty _ = error "Property expected"

File: Haskell/FP-3-uzd/FP3/MExpression/Parse.cs
using System;
using System.Collections.Generic;

namespace FP3
{
    public class Parser
    {
        public static List<Move> Parse(string str)
        {
            string str1 = RemoveWhiteSpaces(str);
            string str2 = RemoveMExpressionPrefix(str1);
            string str3 = RemoveOpeningSquareBracket(str2);
            MovesString movesString = ParseMExpressions(new List<Move>(), str3);
            return movesString.moves;
        }

        private static MovesString ParseMExpressions(List<Move> acc, string mExpression)
        {
            if (mExpression[0] == ']')
                return new MovesString(acc, mExpression);
            if (mExpression[0] == ';')
                return ParseMExpressions(acc, mExpression.Substring(1));

            CharString charString1 = ReadPropertyName(mExpression);
            string mExpression1 = RemoveSemicolon(charString1.str);
            MoveString moveString1 = ParseMExpression(mExpression1);

            acc.Add(moveString1.move);

            return ParseMExpressions(acc, moveString1.str);
        }

        private static MoveString ParseMExpression(string mExpression)
        {
            string mExpression1 = RemoveMExpressionPrefix(mExpression);
            string mExpression2 = RemoveOpeningSquareBracket(mExpression1);

            CharString charString1 = ReadPropertyName(mExpression2);
            string mExpression4 = RemoveSemicolon(charString1.str);
            IntString charString2 = ReadProperty(mExpression4);
            string mExpression6 = RemoveSemicolon(charString2.str);

            CharString charString3 = ReadPropertyName(mExpression6);
            string mExpression8 = RemoveSemicolon(charString3.str);
            IntString charString4 = ReadProperty(mExpression8);
            string mExpression10 = RemoveSemicolon(charString4.str);

            CharString charString5 = ReadPropertyName(mExpression10);
            string mExpression12 = RemoveSemicolon(charString5.str);
            IntString charString6 = ReadProperty(mExpression12);
            string rest = RemoveClosingSquareBracket(charString6.str);

            int x, y, v;
            char trueV;

            if (charString1.c == 'x')
                x = charString2.i;
            else if (charString3.c == 'x')
                x = charString4.i;
            else x = charString6.i;

            if (charString1.c == 'y')
                y = charString2.i;
            else if (charString3.c == 'y')
                y = charString4.i;
            else y = charString6.i;

            if (charString1.c == 'v')
                v = charString2.i;
            else if (charString3.c == 'v')
                v = charString4.i;
            else v = charString6.i;

            trueV = (v == 3) ? 'x' : 'o';

            return new MoveString(new Move(x, y, trueV), rest);
        }

        private static string RemoveWhiteSpaces(string message)
        {
            return RemoveWhiteSpaces2("", message);
        }

        private static string RemoveWhiteSpaces2(string acc, string message)
        {
            if (message == "")
                return acc;
            if (message[0] != ' ')
                return RemoveWhiteSpaces2(acc + message[0], message.Substring(1));
            else
                return RemoveWhiteSpaces2(acc, message.Substring(1));
        }

        private static string RemoveOpeningSquareBracket(string expression)
        {
            if (expression[0] == '[')
                return expression.Substring(1);
            throw new Exception("'[' expected");
        }

        private static string RemoveClosingSquareBracket(string expression)
        {
            if (expression[0] == ']')
                return expression.Substring(1);
            throw new Exception("']' expected");
        }

        private static string RemoveMExpressionPrefix(string expression)
        {
            if (expression[0] == 'm')
                return expression.Substring(1);
            throw new Exception("'m' expected");
        }

        private static string RemoveSemicolon(string expression)
        {
            if (expression[0] == ';')
                return expression.Substring(1);
            throw new Exception("';' expected");
        }

        private static CharString ReadPropertyName(string expression)
        {
            if (expression[0] == '\"' && expression[2] == '\"')
            {
               return new CharString(expression[1], expression.Substring(3));
            }
            throw new Exception("Property name expected");
        }

        private static IntString ReadProperty(string expression)
        {
            if (expression[0] == '0')
                return new IntString(0, expression.Substring(1));
            else if (expression[0] == '1')
                return new IntString(1, expression.Substring(1));
            else if (expression[0] == '2')
                return new IntString(2, expression.Substring(1));
            else if (expression[0] == '\"')
            {
                if (expression[1] == 'x' || expression[1] == 'X')
                {
                    if (expression[2] == '\"')
                        return new IntString(3, expression.Substring(3));
                }
                else if (expression[1] == 'o' || expression[1] == 'O')
                    if (expression[2] == '\"')
                    {
                        return new IntString(4, expression.Substring(3));
                    }
            }
            throw new Exception("Property expected");
        }
    }
}


